proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+
+* To disambiguate the namespaces within one module, a `-XModuleAliases` extension
+  is added that introduces new syntax:
+
+  ```
+  module   -> module modid [exports] where body
+            | body
+  ```
+  is changed to
+  ```
+  module   -> module modid [alias_1, ..., alias_n] [exports] where body
+            | body
+
+  alias -> [namespace] as modid
+
+  namespace -> data
+             | type
+  ```
+    * When `as modid` is specified, the users can refer to the module within
+      the module itself using the provided qualifier.
+    * When `data as modid` is specified, the users can refer to the terms
+      in the module within the module itself using the provided qualifier.
+    * When `type as modid` is specified, the users can refer to the types
+      in the module within the module itself using the provided qualifier.
+
+* Deprecate the `'` syntax of `-XDataKinds`, reserving this syntax for Template
+  Haskell name quotation. Introduce a new warning, `-Wticked-promoted-constructors` which
+  warns of the usage of the syntax.
+
+* Deprecate the `''` syntax in Template Haskell. Introduce a new warning,
+  `-Wdouble-tick-template-haskell` which warns about the usage of the syntax.
+
+* Introduce a new warning, `-Wpun-bindings`, triggered by any name
+  binding that would clash with another identifier if Haskell had a single
+  unified namespace.
+
+* Introduce a new warning, `-Wpuns`, triggered by using an identifier
+  that would be ambiguous or refer to another entity if Haskell had a single
+  unified namespace.
+
+* The deprecation strategy for the `'` syntax in `-XDataKinds` and `''` syntax in `-XTemplateHaskell` is the following:
+  * In the next release add `-Wpun-bindings`, `-Wpuns`, `-Wticked-promoted-constructors` and `-Wdouble-tick-template-haskell` to `-Wcompat` and deprecate `-Wunticked-promoted-constructors`.
+  * Three releases from after this proposal is implemented add all four warnings to `-Wall`.
+  * Seven releases from after this proposal is implemented deprecate the syntax and enable `-Wticked-promoted-constructors` and `-Wdouble-tick-template-haskell` by default.
+  * Fifteen releases from after this proposal is implemented remove the syntax.
+
+* Amend [GHC Proposal #65](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0065-type-infix.rst)
+  to use `data` instead of `value`.
+
+* Under `-XPatternSynonyms`, change the `pattern` qualifier in import and export lists to `data`.
+  Introduce `-Wpattern-namespace-qualifier` warning that warns when the
+  `pattern` namespace qualifier is used.  Add it to `-Wcompat`.
```

Comment:
And then what? Is there a plan to remove support for `pattern` here?

Reply:
Should there be a plan to remove `pattern`? Or do we just keep it deprecated indefinitely? (It's not really causing much harm... but I'd still want this warning to be added to `-Wall` in a few years.)

Reply:
Added that to the proposal!